[Feature] 리뷰 등록, 삭제, 수정 구현

ghrltjdtprbs opened the pull request:
## [Feature] 리뷰 등록, 삭제, 수정 구현

### 리뷰 등록 구현
 - 리뷰 작성 구현.
 - 리뷰 제목, 내용, 긍정태그, 부정태그, 별점 작성 가능.
 - 엔드포인트 : POST api/v2/reviews
 - 리뷰 삭제 후 재등록시 요청 request에 맞춰 복원하도록 예외처리 추
 
### 리뷰 삭제 구현
 - 리뷰 삭제 구현.
 - soft delete 적용
 - 엔드포인트 : DELETE api/v2/reviews/{reviewId}
 - 
### 리뷰 수정구현
 - 리뷰 수정구현.
 - 리뷰 제목, 내용, 긍정태그, 부정태그, 별점 수정 가능.
 - 엔드포인트 : POST api/v2/reviews/{reviewId}
 - 본인의 리뷰만 수정할 수 있도록 예외처리 추가


----------------
### 특이 사항
- member 엔터티에 부트캠프 인증을 위한 campCrtfc, bootcamp 필드 추가
- 부트캠프 이름은 admin이 부여해야 함
- 테스트 완료

ghrltjdtprbs committed:
[feat] 리뷰작성 구현

변경 사항 :
 - 리뷰 작성 구현.
 - 부트캠프 이름, 제목, 내용, 긍정태그, 부정태그, 별점 작성 가능.

ghrltjdtprbs committed:
[refactor] SecurityConfig PERMIT_URL_ARRAY 수정

문제 사항 :
 - member 도메인 엔드포인트 변경으로 PERMIT_URL_ARRAY 수정 필요
변경 사항 :
 - 엔드포인트 변경사항 반영

ghrltjdtprbs committed:
[feat] 리뷰 삭제 구현

변경 사항 :
 - 리뷰 삭제 기능 구현.
 - 엔드포인트 : DELETE api/v2/reviews/{reviewId}
 - soft delete 적용

ghrltjdtprbs committed:
[feat] member엔터티에 campCrtfc 추가

변경 사항 :
 - member엔터티에 캠프 인증 여부를 나타내는campCrtfc 추가
 - default = false

ghrltjdtprbs committed:
[feat] 부트캠프 인증 예외처리 추가

변경 사항 :
 - 리뷰 작성/삭제시 부트캠프 인증을 하지 않았거나 권한이 없을 때의 예외 처리 추가

ghrltjdtprbs committed:
[feat] 리뷰 수정 기능 구현

변경 사항 :
 - 리뷰 수정 기능 구현하였습니다.
 - 엔드포인트 PUT api/v2/reviews/{reviewId}

ghrltjdtprbs committed:
[feat] MEMBER가 수강한 부트캠프 이름을 가지도록 수정

문제 사항 :
 - 리뷰를 작성자가 부트캠프 이름을 직접 입력할 경우 인증로직이 무용지물이 됨
변경 사항 :
 - 각각의 MEMBER가 수강한 부트캠프 이름을 가질 수 있도록 bootcamp 필드 추가
 - 관련 로직 수정

ghrltjdtprbs committed:
[feat] 리뷰 중복 작성 예외처리 추가

문제 사항 :
 - 부트캠프 인증을 한 회원이 한 부트캠프에 대해 여러개의 리뷰를 작성하면 안됨
변경 사항 :
 - ReviewAlreadyExistsException예외 추가

ghrltjdtprbs committed:
[feat] member 엔터티 변경사항 flyway 스크립트 추가

문제 사항 :
 - member 엔터티에 campCrtfc,bootcamp 필드 추가
변경 사항 :
 - 해당 변경사항 flyway 스크립트 작성, 추가
    -> V14__Add_Column_memeber_table.sql

chimaek commented on src/main/java/com/fasttime/domain/member/entity/Member.java:
@Setter가 있어용.

ghrltjdtprbs replied:
수정했습니다! 58367f91c438eddc665ffcc1047d2e5002ac7eb6 커밋 확인해주세요ㅎㅎ

chimaek commented on src/main/java/com/fasttime/domain/member/entity/Member.java:
```diff
@@ -36,6 +36,10 @@ public class Member extends BaseTimeEntity  {
 
     private String nickname;
 
+    private boolean campCrtfc = false;
```
@Builder.Default로 초기화하는 방식도 고려해보세용

ghrltjdtprbs replied:
b74c205195589ed06192b4bfa7e032b7f1480407 확인해주세요! 이렇게 하는게 맞나 모르겠네요.. 😭😭

chimaek commented on src/main/java/com/fasttime/domain/review/controller/ReviewController.java:
각 메서드 부분의 매핑 부분에서 URL이 중복되는데영

@RequestMapping으로 중복을 최소화 해보죠!

ghrltjdtprbs replied:
07e00c40587ff7133ef3a40c6bf06b4eeba83504 확인해주세요!

chimaek commented on src/main/java/com/fasttime/domain/review/controller/ReviewController.java:
```diff
@@ -0,0 +1,61 @@
+            requestDTO.badtags(),
+            reviewService.getTagRepository()
+        );
+        return ResponseEntity.ok(ResponseDTO.res(HttpStatus.CREATED,"리뷰작성이 완료되었습니다.", responseDTO));
```
`return ResponseEntity.status(HttpStatus.CREATED).body(ResponseDTO.res(HttpStatus.CREATED, "리뷰작성이 완료되었습니다.", responseDTO));`

이게 RESTful에 적절합니다.

ghrltjdtprbs replied:
반영했습니다! 2d6c4cc27e8f73f3f1c6edb2c8f8ac9f4d421269 확인해주세요 ㅎㅎ

chimaek commented on src/main/java/com/fasttime/domain/review/controller/ReviewController.java:
```"리뷰작성이 완료되었습니다.", "리뷰삭제가 완료되었습니다.", "리뷰수정이 완료되었습니다."```

이렇게 하드코딩하는 것보다는 상수로 정의해보는 건 어떨까요?

ghrltjdtprbs replied:
우선 성공 메시지를 "리뷰 요청이 완료되었습니다." 로 통일해 상수로 정의했습니다! aef2042d315687dd2edc5d24ffca66fb032969f7 확인해주세요! 만약 필요하다면 enum 클래스로 따로 관리 하는 것이 좋겠네요! 좋은 의견 감사합니다!!  👍👍

chimaek commented on src/main/java/com/fasttime/domain/review/exception/ReviewAlreadyDeletedException.java:
JavaDoc 주석을 넣어보죠 앞으로는 유지,보수성을 위해 넣는것이 좋겠습니다.

chimaek commented on src/main/java/com/fasttime/domain/review/service/ReviewService.java:
```diff
@@ -0,0 +1,101 @@
+        Review review = reviewRepository.findById(reviewId)
+            .orElseThrow(() -> new ReviewNotFoundException());
+
+        review.setupdate(requestDTO.title(),requestDTO.rating(),requestDTO.content());
```
setupdate 네이밍 확인요

ghrltjdtprbs replied:
자바 네이밍 컨벤션에 따라 변경하였습니다! (setupdate -> updateReviewDetails) 868b0b1f66e302de5227fb984c34f7bc34636850 확인해주세요! 😊😊

chimaek commented on src/main/java/com/fasttime/domain/review/dto/response/ReviewResponseDTO.java:
```diff
@@ -0,0 +1,41 @@
+
+    private static Set<String> getTagContents(Set<Long> tagIds, TagRepository tagRepository) {
+        return tagIds.stream()
+            .map(tagId -> tagRepository.findById(tagId)
```
tagRepository를 dto내부에 있어선 될까요? 외부 레이어 단에서 데이터를 만들고(getTagContents와 같은 함수를 서비스 레이어에 위치한다던지?) dto에 전달하는것이 효율있어보입니다.

ghrltjdtprbs replied:
맞습니다! dto 전체적으로 수정했습니다. acc5f1c78d0ae7b1df43c053c3a4b5185a079e68 확인해주세요👍👍

chimaek commented on src/main/java/com/fasttime/domain/review/service/ReviewService.java:
```diff
@@ -0,0 +1,101 @@
+        return reviewTags;
+    }
+
+    public TagRepository getTagRepository() {
+        return this.tagRepository;
```
확인바래요 ㅎㅎ 굳이 이 메서드가 필요할까요?

ghrltjdtprbs replied:
이 부분은 제가 놓쳤습니다.. 메서드 삭제했으니 확인해주세요! acc5f1c78d0ae7b1df43c053c3a4b5185a079e68

chimaek reviewed (commented):
리뷰 완료 고생 많이했어요~

JeongUijeong commented on src/main/java/com/fasttime/domain/review/entity/ReviewTag.java:
```diff
@@ -6,10 +6,14 @@
 
 @Entity
 @Getter
+@AllArgsConstructor
```
빌더 패턴을 쓰기로 했던 것 같은데 혹시 필드가 3개 뿐이라서 의도적으로 AllArgeConstructor 애노테이션을 사용하신 걸까요? 😊

ghrltjdtprbs replied:
해당 테이블은 중간 테이블이라 빌더 패턴을 적용하지 않는 것이 더 효율적이라고 판단했습니다! 요청하신다면 추가하겠습니다 ㅎㅎ 👍👍

JeongUijeong commented on src/main/java/com/fasttime/domain/review/service/ReviewService.java:
```diff
@@ -0,0 +1,101 @@
+
+    public Review createReview(ReviewRequestDTO requestDTO, Long memberId) {
+        Member member = memberRepository.findById(memberId)
+            .orElseThrow(() -> new MemberNotFoundException());
```
알고 계시겠지만 혹시 도움이 될까 적어봅니다. 이렇게 표현하는 것도 방법도 있으니, 참고만 해주세요~😊
```suggestion
            .orElseThrow(MemberNotFoundException::new);
```

ghrltjdtprbs replied:
확인했습니다! 알려주셔서 감사합니다 ㅎㅎ 😊😊

JeongUijeong reviewed (changes requested):
수고하셨습니다!👍🏻 
저희 TDD 방식으로 개발하기로 했었는데, 테스트 코드가 보이지 않네요..!
구현이 먼저 되긴 했지만, 어쨋든 테스트 코드도 추가로 작성해주시면 좋을 것 같습니다~
또, 몇 가지 코멘트 적어두었으니 확인 부탁 드립니다~~😊

ghrltjdtprbs committed:
[refactor] member 엔터티 @Setter 삭제

문제 사항 :
 - member 엔터티에 @Setter 존재
변경 사항 :
 - @Setter 삭제
 - 관련 로직 변경

ghrltjdtprbs committed:
[refactor] member 엔터티 @Builder.Default 추가

변경 사항 :
 - Member의 campCrtfc 필드에 기본값 처리를 위해 @Builder.Default 어노테이션 추가

ghrltjdtprbs committed:
[refactor] ReviewController url 중복 리팩토링

문제 사항 :
 - url 중복
변경 사항 :
 - @RequestMapping 으로 url 중복 최소화

ghrltjdtprbs committed:
[refactor] setupdate 메서드 이름 변경

변경 사항 :
 - setupdate 메서드는 이름을 자바 네이밍 컨벤션에 따라 변경
   setupdate -> updateReviewDetails
 - 이름 변경으로 관련 로직 수정

ghrltjdtprbs committed:
[refactor] 리뷰 생성 메서드 http 응답 상태 코드 수정

변경 사항 :
 - 리뷰 생성 메서드 http 응답 상태 코드를 RESTful 하게 수정

ghrltjdtprbs committed:
Merge branch 'develop' of https://github.com/fcstudy-project/FastTime-backend into feature/review

ghrltjdtprbs committed:
[fix] MemberControllerTest 수정

문제 사항 :
 - Member 엔터티의 @Setter 삭제로 기존 테스트 코드 실패
변경 사항 :
 - 기존 set 을 빌더로 변경

ghrltjdtprbs committed:
[refactor] ReviewResponseDTO 리팩토링

문제 사항 :
 - ReviewResponseDTO에 tagRepository 존재
변경 사항 :
 - 해당 기능을 서비스 단으로 옮김
 - 본인이 작성한 리뷰만 수정할 수 있도록 예외처리 추가

ghrltjdtprbs committed:
[refactor] 리뷰 삭제후 리뷰 재등록 예외추가

문제 사항 :
 - 리뷰를 삭제했음에도 이미 리뷰를 작성했다는 예외로 잡히는 문제
변경 사항 :
 - 리뷰 복원기능 추가
 - 리뷰 삭제 후 리뷰를 재작성하면 soft delete된 기존 리뷰 복원과 동시에 재등록시 보낸 request 값으로 리뷰 수정

ghrltjdtprbs committed:
[refactor] 성공 메시지 상수로 정의

문제 사항 :
 - 성공 메시지를 하드코딩 하는 것보다 상수로 정의하라는 피드백 수용
변경 사항 :
 - REVIEW_SUCCESS_MESSAGE 로 성공 메시지 통일

ghrltjdtprbs commented:
> 리뷰 완료 고생 많이했어요~

피드백 전부 반영해 커밋했습니다! 확인해주세요 ㅎㅎ

ghrltjdtprbs commented:
> 수고하셨습니다!👍🏻 저희 TDD 방식으로 개발하기로 했었는데, 테스트 코드가 보이지 않네요..! 구현이 먼저 되긴 했지만, 어쨋든 테스트 코드도 추가로 작성해주시면 좋을 것 같습니다~ 또, 몇 가지 코멘트 적어두었으니 확인 부탁 드립니다~~😊

테스트 코드는 다음 pr에 작성해 올리겠습니다! 감사합니다 ㅎㅎ 😊😊😊

JeongUijeong reviewed (approved):
수고하셨습니다~!😊👍🏻❤️
